src/dirent_lookup.h

```diff
@@ -23,59 +23,95 @@
 #include "zim_types.h"
 #include "debug.h"
 #include "narrowdown.h"
+#include "_dirent.h"
 
 #include <algorithm>
 #include <map>
 #include <mutex>
 #include <vector>
+#include <cassert>
 
 namespace zim
 {
 
-template<class Impl>
+template<class TConfig>
 class DirentLookup
 {
 public: // types
+  typedef typename TConfig::DirentAccessorType DirentAccessor;
   typedef std::pair<bool, entry_index_t> Result;
 
 public: // functions
-  DirentLookup(const Impl* _impl, entry_index_type cacheEntryCount);
+  explicit DirentLookup(const DirentAccessor* _direntAccessor);
 
   entry_index_t getNamespaceRangeBegin(char ns) const;
   entry_index_t getNamespaceRangeEnd(char ns) const;
 
-  Result find(char ns, const std::string& url);
+  Result find(char ns, const std::string& key) const;
 
-private: // functions
-  std::string getDirentKey(entry_index_type i) const;
+protected: // functions
+  int compareWithDirentAt(char ns, const std::string& key, entry_index_type i) const;
+  Result findInRange(entry_index_type l, entry_index_type u, char ns, const std::string& key) const;
 
-private: // types
+protected: // types
   typedef std::map<char, entry_index_t> NamespaceBoundaryCache;
 
-private: // data
-  const Impl* impl = nullptr;
+protected: // data
+  const DirentAccessor& direntAccessor;
+  const entry_index_type direntCount;
 
   mutable NamespaceBoundaryCache namespaceBoundaryCache;
   mutable std::mutex cacheAccessMutex;
+};
+
+template<class TConfig>
+int DirentLookup<TConfig>::compareWithDirentAt(char ns, const std::string& key, entry_index_type i) const
```

Comment:
I always confused about what is returned by compare string which return a int.
It would be nice to have small unit test for this method.

Reply:
Done